### PR TITLE
Add tests for SOCKS proxy support

### PR DESCRIPTION
**Pull request by guillon**
This pull request adds tests/test_socks.cpp for testing socks proxy support.

Please do not merge yet, until I can check if the coverage report is what I should expect.

Also I added the test to Linux only for now as I do not have the possibility to test on other hosts.

### CODE REVIEW

File: tests/test_socks.cpp [hunk 1]
```diff
@@ -0,0 +1,564 @@
+/*
+    Copyright (c) 2007-2018 Contributors as noted in the AUTHORS file
```
Comment (bluca) [inline tests/test_socks.cpp]: 2019

**Comment by bluca** [comment]
I've merged the other PR, please rebase to remove the formatting commit when you have a moment. Other than that, this is great, thanks!

File: tests/test_socks.cpp (hunk 1 above)
Comment (guillon) [reply tests/test_socks.cpp]: ok, will fix the year.

**Comment by guillon** [comment]
> I've merged the other PR, please rebase to remove the formatting commit when you have a moment. Other than that, this is great, thanks!
Yes, of course, will rebase and check results.

**Comment by bluca** [comment]
The test seems to be failing on Travis:

```
FAIL: tests/test_socks
======================
Push server bound to: tcp://127.0.0.1:45612
Pull client connected to: tcp://127.0.0.1:45612
remote: sending 2 messages
client: receiving 2 messages
socks_server: starting server thread
Push server bound to: tcp://127.0.0.1:32853
Pull client connected to: tcp://127.0.0.1:32853
remote: sending 2 messages
socks_server: listening at address: 127.0.0.1:46387
socks_server: accepted client connection 1/1: client fd 15
socks_server: received greetings
socks_server: answered greetings (method: 0x0)
socks_server: received command (cmd: 1)
socks_server: received address (addr: 0x7f000001)
socks_server: received port (port: 32853)
socks_server: trying to connect to 7f000001:32853
socks_server: replied to request (err: 0x0)
socks_server: waiting for input (client fd: 15, remote fd: 16)
client: receiving 2 messages
socks_server: ready to read from fd 16
socks_server: ready to read from fd 15
socks_server: ready to read from fd 16
socks_server: ready to read from fd 15
socks_server: ready to read from fd 15
socks_server: ready to read from fd 16
socks_server: ready to read from fd 16
socks_server: ready to read from fd 15
socks_server: closed remote connection 1/1: client fd -1
socks_server: closed client connection 1/1: client fd -1
socks_server: closed server
Push server bound to: tcp://127.0.0.1:42580
Pull client connected to: tcp://127.0.0.1:42580
remote: sending 2 messages
socks_server: starting server thread
socks_server: listening at address: 127.0.0.1:42224
```

**Comment by bluca** [comment]
it's a timeout, so it's taking more than 60 seconds

**Comment by guillon** [comment]
> it's a timeout, so it's taking more than 60 seconds

Yes, normally, it's very quick, hence I guess that the subscriber can't connect to the proxy or the proxy to the publisher. May be an issue due to timing, I will check locally.

**Comment by guillon** [comment]
Ok, it's an issue with the startup order (depends upon the host load) of the proxy server versus the zmp pull client.
I will add some semaphore in order to control the startup sequence.

Note that I published pr #3542 as I had to fix the assertion report in order to isolate the right line number where the test failed.

**Comment by bluca** [comment]
The connection should be retried automatically if the endpoint is not reachable, so isn't this highlighting a bug in the implementation?

**Comment by guillon** [comment]
You're right, the problem actually is not the connection hanshake but simply the server bind address which is not yet determined (because we ask for an arbitrary port).
It's a bug in the test. I will fix it, actually I do not need any timing nor semaphore, I just have to bind the sockets first, before launching the proxy thread.

**Comment by guillon** [comment]
Fixed the issue in the test hopefully. The actual threads ordering/timing should not be a problem anymore.

Will wait for the coverage report and see if I can improve it...

**Comment by bluca** [comment]
looks good - seems like the main part missing from coverage is ipv6, which is not easy to get on travis unfortunately

**Comment by guillon** [comment]
I think that I can improve the coverage, please wait before merging.

Note that the appveyor tests do not fail due to this patch.

**Comment by bluca** [comment]
Sure - don't worry about appveyor

**Comment by guillon** [comment]
I completed the socks tests to a reasonable level of coverage for the three related files options.cpp/socks_connecter.cpp/socks.cpp.
I guess that the coverage should be maximized except for the the case of malformed server error (calls to error() in the code).

You can merge it if all is good. Thanks for your help.

**Pull request merged**